src/data/option/basic.lean

```diff
@@ -105,6 +105,31 @@ by cases x; simp
 
 @[simp] theorem map_id' : option.map (@id α) = id := map_id
 
+/--
+Composing an `option.map` with another `option.map` is equal to
+a single `option.map` of a composition of functions, fully applied.
+This is the reverse direction of `option.comp_map`.
+-/
+@[simp] lemma map_map (h : β → γ) (g : α → β) (x : option α) :
+  option.map h (option.map g x) = option.map (h ∘ g) x :=
+by { cases x; simp only [map_none', map_some'] }
```

Comment:
FWIW @jcommelin, I think this is consistent with `list.map_map`

Reply:
Should `simp` be reducing the number of terms? Then this is the direction, just like `list.map_map` or `prod.map_map`.

Reply:
Reducing the number of terms is a good rule of thumb. Here, I would add that `map_map` puts `h` and `g` next to each other so if we have `simp` rules about `h` and `g` or their constituents they can get triggered after `map_map` and continue the simplification further.

Reply:
If the argument is to enable further simplification, would `option.map (λ a, h $ g a) x` be a better lemma? With `option.map (h ∘ g) x`, it's not clear to me how to unfold further - `simp [function.comp]` doesn't work, does it?

Reply:
If I understand correctly, that lambda expression is fully equivalent to the composition expression because `function.comp` is inlined and reducible. @cipher1024 where do you mean that `h` ang `g` should be next to each other? That lambda expression style?

In fact, the composition results in better `simp` performance, in the following test:

```
import data.option.basic

open option

namespace option

/--
Composing an `option.map` with another `option.map` is equal to
a single `option.map` of a composition of functions, fully applied.
This is the reverse direction of `option.comp_map`.
-/
lemma map_map {α β γ : Type*}
  (h : β → γ) (g : α → β) (x : option α) :
  option.map h (option.map g x) = option.map (h ∘ g) x :=
by { cases x; simp only [map_none', map_some'] }

lemma map_map' {α β γ : Type*}
  (h : β → γ) (g : α → β) (x : option α) :
  option.map h (option.map g x) = option.map (λ a, h (g a)) x :=
by { cases x; simp only [option.map_none', option.map_some'] }

attribute [simp] function.left_inverse.id

example {α β : Type*} (h : β → α) (g : α → β) (hi : function.left_inverse h g) (x : option α) :
  option.map h (option.map g x) = x :=
by { simp [map_map, hi] } -- solved

example {α β : Type*} (h : β → α) (g : α → β) (hi : function.left_inverse h g) (x : option α) :
  option.map h (option.map g x) = x :=
by { simp [map_map', hi] } -- not solved

end option
```

Reply:
`map_map` is more standard, I would stick with it. Then the formalization of `g` and `h` can be tailored to the result of using `map_map`. You can have a look at the [point-free lemmas](https://github.com/leanprover-community/mathlib/blob/27f622eff9ab56d2fe7f8f619dc5f6b4caf1c31e/src/control/bifunctor.lean#L46) in bifunctors. Sometimes a point-free format (i.e. `h ∘ g` instead of `h (g x)`) is necessary, sometimes point-full is. The `higher_order` attribute generates point-free formulations for a given lemma. It's useful to have both.